node/reqres.js

```diff
@@ -86,23 +106,39 @@ function TChannelOutgoingResponse(id, options, sendFrame) {
     }
     options = options || {};
     var self = this;
+    EventEmitter.call(self);
     self.id = id || 0;
     self.code = options.code || 0;
     self.tracing = options.tracing || emptyTracing;
     self.headers = options.headers || {};
     self.checksumType = options.checksumType || 0;
+    self.ok = true;
     self.name = options.name || '';
     self.arg2 = options.arg2 || emptyBuffer;
     self.arg3 = options.arg3 || emptyBuffer;
     self.sendFrame = sendFrame;
 }
 
+inherits(TChannelOutgoingResponse, EventEmitter);
+
 TChannelOutgoingResponse.prototype.send = function send(err, res1, res2) {
     var self = this;
-    self.sendFrame(err, res1, res2);
+    if (err) {
+        self.ok = false;
+        var errArg = isError(err) ? err.message : JSON.stringify(err); // TODO: better
+        self.sendFrame(self.name, res1, errArg);
+    } else {
+        self.sendFrame(self.name, res1, res2);
+    }
 };
 
 module.exports.IncomingRequest = TChannelIncomingRequest;
 module.exports.IncomingResponse = TChannelIncomingResponse;
 module.exports.OutgoingRequest = TChannelOutgoingRequest;
 module.exports.OutgoingResponse = TChannelOutgoingResponse;
+
+function isError(obj) {
+    return typeof obj === 'object' && (
+        Object.prototype.toString.call(obj) === '[object Error]' ||
+        obj instanceof Error);
+}
```

Comment:
The pervasive replication of this non-normalized logic needs to stop. I swear to you on a stack of Silmarillions, `Object.prototype.toString.call(obj) === '[object Error]'` is equivalent and sufficient for all values of `obj` including `null`, `undefined`, numbers, strings, custom objects, and instances of `Error` from other JavaScript contexts, unless someone decides to subvert Object.prototype.toString, which one would only ever do to deliberately alter the behavior of this function.